module/zfs/dsl_scan.c

```diff
@@ -1545,14 +2468,331 @@ dsl_scan_visit(dsl_scan_t *scn, dmu_tx_t *tx)
+	ASSERT0(queue->q_issuing_start);
+	ASSERT0(queue->q_issuing_end);
+
+	/* handle tunable overrides */
+	if (scn->scn_checkpointing || scn->scn_clearing) {
+		if (zfs_scan_issue_strategy == 1) {
+			return (range_tree_first(queue->q_exts_by_addr));
+		} else if (zfs_scan_issue_strategy == 2) {
+			return (avl_first(&queue->q_exts_by_size));
+		}
+	}
+
+	/*
+	 * During normal clearing, we want to issue our largest segments
+	 * first, keeping IO as sequential as possible, and leaving the
+	 * smaller extents for later with the hope that they might eventually
+	 * grow to larger sequential segments. However, when the scan is
+	 * checkpointing, no new extents will be added to the sorting queue,
+	 * so the way we are sorted now is as good as it will ever get.
+	 * In this case, we instead switch to issuing extents in LBA order.
+	 */
+	if (scn->scn_checkpointing) {
+		return (range_tree_first(queue->q_exts_by_addr));
+	} else if (scn->scn_clearing) {
+		return (avl_first(&queue->q_exts_by_size));
+	} else {
+		return (NULL);
+	}
+}
+
+static void
+scan_io_queues_run_one(void *arg)
+{
+	dsl_scan_io_queue_t *queue = arg;
+	kmutex_t *q_lock = &queue->q_vd->vdev_scan_io_queue_lock;
+	scan_io_t *sio = NULL;
+	list_t sio_list;
+	const range_seg_t *rs;
+	boolean_t suspended = B_FALSE;
+
+	ASSERT(queue->q_scn->scn_is_sorted);
+
+	list_create(&sio_list, sizeof (scan_io_t),
+	    offsetof(scan_io_t, sio_nodes.sio_list_node));
+	mutex_enter(q_lock);
+
+	/* reset per-queue scan statistics for this txg */
+	queue->q_avg_seg_size_this_txg = 0;
+	queue->q_segs_this_txg = 0;
+	queue->q_avg_zio_size_this_txg = 0;
+	queue->q_zios_this_txg = 0;
+
+	/* loop until we have run out of time or zios */
+	while ((rs = scan_io_queue_fetch_ext(queue)) != NULL) {
+		scan_io_t *first_io, *last_io;
+
+		/*
+		 * We have selected which extent needs to be processed next,
+		 * gather up the corresponding zio's.
+		 */
+		scan_io_queue_gather(queue, rs, &sio_list);
+
+		first_io = list_head(&sio_list);
+		last_io = list_tail(&sio_list);
+		if (first_io == NULL)
+			continue;
+
+		ASSERT(first_io != NULL);
+		ASSERT(last_io != NULL);
+
+		/*
+		 * We have zio's to issue. Construct a fake range
+		 * seg that covers the whole list of zio's to issue
+		 * (the list is guaranteed to be LBA-ordered) and
+		 * save that in the queue's "in flight" segment.
+		 * This is used to prevent freeing I/O from hitting
+		 * that range while we're working on it.
+		 */
+		queue->q_issuing_start = SCAN_IO_GET_OFFSET(first_io);
+		queue->q_issuing_end =
+		    SCAN_IO_GET_OFFSET(last_io) + last_io->sio_asize;
+
+
+		scan_io_queues_update_seg_stats(queue,
+		    queue->q_issuing_start, queue->q_issuing_end);
+
+		/* issuing zio's can take a long time so drop the queue lock. */
+		mutex_exit(q_lock);
+		suspended = scan_io_queue_issue(queue, &sio_list);
+		mutex_enter(q_lock);
+
+		/* invalidate the in-flight I/O range */
+		queue->q_issuing_start = 0;
+		queue->q_issuing_end = 0;
```

Comment:
it looks like scan_io_queue_issue does not wait for the i/os to complete, so it's wrong to reset q_issuing_* here, before they've completed.